File: python/2015/day08b/day08b.py
#!/usr/bin/env python3

import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    result = 0
    with open(filename, 'r') as infile:
        for line in infile:
            code_len = len(line)
            enc_len = 0
            for i in range(len(line)):
                if line[i] == '\\' or line[i] == '"':
                    enc_len += 2
                else:
                    enc_len += 1
            result += 2 + (enc_len - code_len)
    return result

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day08b/day08b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my $result = 0;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        my $codeLen = length($line);
        my $encLen = 0;
        for my $i (0..(length($line) - 1)) {
            if ( substr($line, $i, 1) eq '\\' || substr($line, $i, 1) eq '"' ) {
                $encLen += 2;
            } else {
                $encLen += 1;
            }
        }
        $result += 2 + ($encLen - $codeLen);
    }
    close($infile);

    return $result;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
